File: app/src/main/java/com/aayush/filmipedia/util/datasource/factory/MovieCrewDataFactory.java
package com.aayush.filmipedia.util.datasource.factory;

import com.aayush.filmipedia.FilmipediaApplication;
import com.aayush.filmipedia.model.MovieCrew;
import com.aayush.filmipedia.util.datasource.MovieCrewDataSource;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import io.reactivex.disposables.CompositeDisposable;

public class MovieCrewDataFactory extends DataSource.Factory<Long, MovieCrew> {
    private MutableLiveData<MovieCrewDataSource> mutableLiveData;
    private MovieCrewDataSource movieCrewDataSource;
    private FilmipediaApplication application;
    private CompositeDisposable compositeDisposable;

    private Long movieId;

    public MovieCrewDataFactory(FilmipediaApplication application,
                                CompositeDisposable compositeDisposable, Long movieId) {
        this.application = application;
        this.compositeDisposable = compositeDisposable;
        this.movieId = movieId;
        mutableLiveData = new MutableLiveData<>();
    }

    @Override
    public DataSource<Long, MovieCrew> create() {
        movieCrewDataSource = new MovieCrewDataSource(application, compositeDisposable, movieId);
        mutableLiveData.postValue(movieCrewDataSource);
        return movieCrewDataSource;
    }

    public MutableLiveData<MovieCrewDataSource> getMutableLiveData() {
        return mutableLiveData;
    }
}

File: app/src/main/java/com/aayush/filmipedia/util/datasource/factory/MovieCrewDataFactory.kt
package com.aayush.filmipedia.util.datasource.factory

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import com.aayush.filmipedia.FilmipediaApplication
import com.aayush.filmipedia.model.MovieCrew
import com.aayush.filmipedia.util.datasource.MovieCrewDataSource
import io.reactivex.disposables.CompositeDisposable

class MovieCrewDataFactory(
    private val application: FilmipediaApplication,
    private val compositeDisposable: CompositeDisposable,
    private val movieId: Long?
): DataSource.Factory<Long, MovieCrew>() {
    val mutableLiveData: MutableLiveData<MovieCrewDataSource> = MutableLiveData()
    private lateinit var movieCrewDataSource: MovieCrewDataSource

    override fun create(): DataSource<Long, MovieCrew> {
        movieCrewDataSource = MovieCrewDataSource(application, compositeDisposable, movieId)
        mutableLiveData.postValue(movieCrewDataSource)
        return movieCrewDataSource
    }
}